# svm

### importing libraries

In [12]:
import pandas as pd 
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC 
from sklearn import metrics

### loading data

In [43]:
df=pd.read_csv('docs_1\DATA.csv') 
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

### splitting the data into x and y to train a model

In [44]:
y=df['label']
x=df.drop('label',axis=1) 
label=LabelEncoder() 
y=label.fit_transform(y) 

### accuracy before standerdization

In [45]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)
model_svm=SVC(gamma='auto') 
model_svm.fit(x_train,y_train) 
y_pred=model_svm.predict(x_test) 
print('accuracy is {}'.format(metrics.accuracy_score(y_pred,y_test)))



accuracy is 0.7192429022082019


### accuracy after standerdization:

#### standardizationis-> mean of zero and a standard deviation of one

In [46]:
scaler=StandardScaler() 
scaler.fit(x) 
x=scaler.transform(x) 
x

array([[-4.04924806,  0.4273553 , -4.22490077, ..., -1.43142165,
        -1.41913712, -1.45477229],
       [-3.84105325,  0.6116695 , -3.99929342, ..., -1.41810716,
        -1.4058184 , -1.01410294],
       [-3.46306647,  1.60384791, -4.09585052, ..., -1.42920257,
        -1.41691733, -1.06534356],
       ...,
       [-1.29877326,  2.32272355, -0.05197279, ..., -0.5992661 ,
        -0.58671739,  0.17588664],
       [-1.2452018 ,  2.012196  , -0.01772849, ..., -0.41286326,
        -0.40025537,  1.14916112],
       [-0.51474626,  2.14765111, -0.07087873, ..., -1.27608595,
        -1.2637521 ,  1.47567886]])

In [47]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)
svm_model=SVC(gamma='auto')
svm_model.fit(x_train,y_train) 
y_predict=svm_model.predict(x_test)
accuracy=metrics.accuracy_score(y_predict,y_test) 
print('accuracy is -> {}'.format(accuracy))

accuracy is -> 0.9763406940063092


#### ACCURACY BEFORE STANDERDIZATION IS :0.719242902208201

#### ACCURACY AFTER STANDERDIZATION IS:0.9763406940063092

###### Standardization refers to shifting the distribution of each attribute to have a mean of zero and a standard deviation of one (unit variance). It is useful to standardize attributes for a model. Standardization of datasets is a common requirement for many machine learning estimators implemented in scikit-learn; they might behave badly if the individual features do not more or less look like standard normally distributed data.

#### Lets try svm with different kernals

In [48]:
#polynomial kernal
svm_model=SVC(kernel='poly',gamma='auto') 
svm_model.fit(x_train,y_train) 
y_pred=svm_model.predict(x_test) 
print('accuracy is -> {}'.format(metrics.accuracy_score(y_pred,y_test)))

accuracy is -> 0.9589905362776026


In [49]:
#rbf kernal
svm_model=SVC(kernel='rbf',gamma='auto')
svm_model.fit(x_train,y_train) 
y_pred=svm_model.predict(x_test) 
print('accuracy is -> {}'.format(metrics.accuracy_score(y_pred,y_test))) 



accuracy is -> 0.9763406940063092


In [50]:
#linear kernal
svm_model=SVC(kernel='linear',gamma='auto')
svm_model.fit(x_train,y_train) 
y_pred=svm_model.predict(x_test) 
print('accuracy is -> {}'.format(metrics.accuracy_score(y_pred,y_test))) 


accuracy is -> 0.9779179810725552


#### from the above data i can canclude that linear_svm is better than polynomial and rbf

### Performing K-fold cross validation with different kernels

#### cv on linear

In [51]:
from sklearn.model_selection import cross_val_score
svm_model=SVC(kernel='linear',gamma='auto')
cross_valid_score=cross_val_score(svm_model,x,y,cv=10,scoring='accuracy') 
print(cross_valid_score)

[0.91194969 0.97169811 0.97169811 0.97169811 0.96202532 0.98734177
 0.99050633 0.98101266 0.95253165 0.99367089]


##### With K-fold cross validation we can see that the dataset got split into 10 equal parts thus covering all the data into training as well into testing set.This is the reason we got 10 different accuracy score.

##### we can see above how the accuracy is different every time.Because this is due to splitting the data randomly 

In [52]:

print('linear score ->{}'.format(cross_valid_score.mean())) 

linear score ->0.9694132632752168


##### we generally take the mean of the scores in k_fold cross validation

#### cv on rbf kernal

In [53]:
from sklearn.model_selection import cross_val_score
svm_model=SVC(kernel='rbf',gamma='auto')
rbf_cross_valid_score=cross_val_score(svm_model,x,y,cv=10,scoring='accuracy') 
print(rbf_cross_valid_score)

[0.93396226 0.95597484 0.96855346 0.9591195  0.96835443 0.99683544
 0.98734177 0.98101266 0.91455696 0.99367089]


In [54]:
print('RBF score ->{}'.format(rbf_cross_valid_score.mean())) 

RBF score ->0.9659382214791815


#### cv on polynomial kernal 

In [55]:
from sklearn.model_selection import cross_val_score
svm_model=SVC(kernel='poly',gamma='auto')
poly_cross_valid_score=cross_val_score(svm_model,x,y,cv=10,scoring='accuracy') 
print(poly_cross_valid_score)

[0.89937107 0.94654088 0.93081761 0.91509434 0.95886076 1.
 0.98101266 0.96835443 0.87974684 0.9778481 ]


In [56]:
print('polynomial score ->{}'.format(poly_cross_valid_score.mean()))

polynomial score ->0.9457646684181196
